### DISCUSSION TITLE
Custom ScrollViewer.Extent does not get set

### DISCUSSION THREAD

**Discussion by MichaelRumpler** [AvaloniaUI/Avalonia#16459] [Q&A] [answered]
I have a custom `ScrollViewer` which sets it's `Content` to a custom `Control`. The `Control` renders itself with SkiaSharp.
Based on my business logic the height of my `Control` increases. I overrode `MeasureOverride` and this returns the new size. This also sets `Canvas.Bounds` to the new value.

Now I want to scroll the `ScrollViewer` to show the new data. But here the `Extent` always remains 0,0. `Viewport` and `Offset` also remain 0,0. Even when I try to set the `Offset` to something different. The latter makes sense. When `Extent` is 0,0 then it can't scroll.

The code looks like this:

	public partial class MyScrollViewer : ScrollViewer, IDisposable
	{
		InnerCanvas Canvas;
		public TerminalView()
		{
			Canvas = new InnerCanvas();
			Content = Canvas;
		}
		
		private void OnSomeBusinessLogic()
		{
			Canvas.InvalidateMeasure();

			Offset = new(0, y);
		}
	}

	internal class InnerCanvas : Control, IDisposable
	{
		protected override Size MeasureOverride(Size availableSize)
		{
			// returns some calculated size which may be bigger than availableSize
		}

		public override void Render(DrawingContext context)
		{
			// render using SKCanvas
		}
	}

What am I doing wrong?

**Comment by IanRawley** [comment]
Extents for a ScrollViewer seem to be set by the ScrollContentPresenter during the Arrange pass.  What control template are you using for your ScrollViewer?  Is it a custom one without a ScrollContentPresenter?  If so, you probably also need to override Arrange and set the extents yourself.

It looks like both the Simple and Fluent themes for ScrollViewer include a ScrollContentPresenter in their templates, but it's not actually required by ScrollViewer like the ScrollBars are.  That's either an oversight, or it needs to be documented somewhere that if your template doesn't use a ScrollContentPresenter you need to handle Extents manually.

Specifically here: https://github.com/AvaloniaUI/Avalonia/blob/d3f234bad58d8b45f13f981e3e1926ee20f30f49/src/Avalonia.Controls/Presenters/ScrollContentPresenter.cs#L726C1-L726C64

**Reply by IanRawley** [reply]
Another issue you might end up having in the example is that while your DoBusinessLogic method has called InvalidateMeasure, that doesn't mean Measure has been recalculated by the time you try to adjust Offset.  You'll probably need to wrap that in a Dispatcher.UIThread.Post()

**Comment by MichaelRumpler** [comment, marked as answer]
Thanks for mentioning the control template. That brought me to the conclusion: I don't use any. I just had a class which inherits from `ScrollViewer`. It was missing

	protected override Type StyleKeyOverride => typeof(ScrollViewer);

Now `Extent` does get set and I can scroll. Thanks!

**Reply by IanRawley** [reply]
A lack of template would certainly do it :)
It's usually a lot more obvious due to not seeing anything at all, but manually rendering parts and handling size masked that.  
I do think it's a little un-intuitive that a ScrollContentPresenter sets the extents of a ScrollViewer, rather than the ScrollViewer observing and setting its own extents appropriately, but that would tie a ScrollViewer directly to ScrollContentPresenter where the current setup leaves the option to skip the ScrollContentPresenter and handle that all yourself.
